ragthoven/__init__.py

```diff
@@ -136,6 +139,150 @@ def __init__(
+            elif inspect.isclass(cfg):
+                cls = cfg
+                # Normalize to the dotted path expected by get_class
+                name = (
+                    f"{cls.__module__.split('ragthoven.tools.', 1)[-1]}.{cls.__name__}"
+                )
+            else:
+                name = cfg
+
+            if name is None:
+                raise ValueError("Tool configuration is missing 'name'")
+
+            if "." not in name:
+                raise ValueError(
+                    "Tool name must include its module prefix, e.g. 'reasoning_tools.Calculator'"
+                )
+
+            if cls is None:
+                cls = get_class("ragthoven.tools", name)
+                class_name = get_class_func_name_only(name)
+            else:
+                class_name = cls.__name__
+
+            requires = getattr(cls, "requires", [])
+            kwargs = {}
+            for dep in requires:
+                if dep not in available_deps:
+                    raise ValueError(f"Unsupported dependency requested: {dep}")
+                kwargs[dep] = available_deps[dep]
+            if model_override is not None:
+                kwargs["model_override"] = model_override
+
+            tools[class_name] = cls(**kwargs)
+
+        return tools
+
+    def _get_tool_schemas(self, tools: dict):
+        return [tool.get_json_schema() for tool in tools.values()]
+
+    def _execute_tool(self, tool_call, tools: dict):
+        try:
+            tool = tools[tool_call.function.name]
+        except KeyError:
+            return f"ERROR: Unknown tool '{tool_call.function.name}'"
+
+        try:
+            args = json.loads(tool_call.function.arguments)
+        except Exception as exc:
+            return f"ERROR: InvalidArguments: {exc}"
+
+        print(f"[RAGTHOVEN][TOOL] call={tool_call.function.name} args={args}")
+
+        try:
+            result = tool(args)
+            print(f"[RAGTHOVEN][TOOL] result={result}")
+
+            # When the ReturnResult tool is invoked, record the final answer so the loop can stop.
+            if tool_call.function.name == ReturnResult.__name__:
+                return result
+
+            return result
+        except Exception as exc:
+            print(f"[RAGTHOVEN][TOOL] error={type(exc).__name__}: {exc}")
+            return f"ERROR: {type(exc).__name__}: {exc}"
+
+    def execute_iterative_loop(self, i, text, all_features):
+        """
+        Iterative execution mode: let the LLM call tools in a loop until it stops.
+        """
+        tool_cfgs = []
+        if self.config.iterative and self.config.iterative.tools:
+            tool_cfgs = list(self.config.iterative.tools)
+        tool_cfgs.append(ReturnResult)
+
+        tools = self._instantiate_tools(tool_cfgs if self.config.iterative else None)
+        max_iter = (
+            self.config.iterative.max_iterations
+            if self.config.iterative is not None
+            else 1
```

Comment:
This logic is slightly confusing - when `self.config.iterative` is None, this passes `None` to `_instantiate_tools()`, but `tool_cfgs` already contains `ReturnResult` by this point. Consider clarifying or adding a comment.